### PR TITLE
Make causal + mask handiling consistent

### PR DESCRIPTION
**Pull request by drisspg**
# Summary
Currently Efficient Attention allows the is_causal flag as well as the attn_mask argument. The math backend explicitly disables this. For causal attention we can iterate over half the number of key/value pairs for fused kernels it is still beneficial to have this marked separately. 

Ideally the mask alone would be able to specify this.

### CODE REVIEW

**Comment by sanchitintel** [comment]
I noticed that even if Attention mask is absent ([it may or may not be present in GPT2](https://github.com/huggingface/transformers/blob/main/src/transformers/models/gpt2/modeling_gpt2.py#L204-L206)), pattern-matching & replacing GPT2 SDPA computation (I can see with a breakpoint that the pattern is matched & the replacement function is run with FakeTensors, but it's not eventually replaced) with [torch.nn.functional.scaled_dot_product_attention](https://github.com/pytorch/pytorch/pull/121570) leads to graph breaks for `hf_GPT2` inference, at least on CPU.

**Comment by sanchitintel** [comment]
Hi! I read your blog-post on  GPT2 performance with the SDPA op, using the HF Optimum library.
I was mistaken about GPT2 not being able to use `is_causal` as `True` for SDPA.
It does use it in [one specific case](https://github.com/huggingface/optimum/blob/fd47a73267c3a71ea4e3c02f92260ae61c5ae372/optimum/bettertransformer/models/attention.py#L58-L60).

However, @valentine233 came up with an approach to use `is_causal` as `False`, [similar to how it's being done in HF Optimum](https://github.com/huggingface/optimum/blob/fd47a73267c3a71ea4e3c02f92260ae61c5ae372/optimum/bettertransformer/models/attention.py#L73-L78).
If we'd follow such an approach, then GPT2 wouldn't need both `is_causal` as `True`, and an attention mask while invoking the SDPA op in a replacement pattern.

But this PR could still benefit some other models.

Thanks!
